go/tools/bazel_testing/bazel_testing.go

```diff
@@ -416,6 +422,46 @@ func setupWorkspace(args Args, files []string) (dir string, cleanup func() error
 		}
 	}
 
+	// If a MODULE.bazel file is requested, create one.
+	if args.ModuleFileSuffix != "" {
+		moduleBazelPath := filepath.Join(mainDir, "MODULE.bazel")
+		if _, err = os.Stat(moduleBazelPath); err == nil {
+			return "", cleanup, fmt.Errorf("ModuleFileSuffix set but MODULE.bazel exists")
+		}
+		var w *os.File
+		w, err = os.Create(moduleBazelPath)
+		if err != nil {
+			return "", cleanup, err
+		}
+		defer func() {
+			if cerr := w.Close(); err == nil && cerr != nil {
+				err = cerr
+			}
+		}()
```

Comment:
This is what I used for such case (you can probably still find this pattern used quite a bit under `src/code.uber.internal/infra/...`, but it was never made official). First, prefix all return variables with `_`:

```go
func setupWorkspace(args Args, files []string) (_dir string, _cleanup func() error, _err error) {
```

That way, it's always clear which value we are referring to: the returned one or the one in the current scope? With that, you are allowed to use the return values *only* in `defer`. Like this:

```suggestion
		defer func() {
			if err := w.Close(); err == nil {
				_err = errors.Join(_err, err)
			}
		}()
```

Then nesting, shadowing works as expected. I didn't have time to spread this pattern, nor create a linter. :)

Reply:
Thanks, I like that pattern. I'll leave this as is for now since we'll likely refactor all of this when we make it ready for running in Bzlmod (not just running a Bzlmod project).